Proposal: multi modules project

Ant-hem opened the pull request:
Split the library into multiple modules

`algoalisearch-core` which is httpClient agnostic
`algoliasearch-apache` which has an implementation of HttpRequester

Ant-hem committed:
Added: personalizationImpact

Ant-hem committed:
Architecture: multi-modules project

aseure committed:
chore: change deploy step to publish to bintray rather than sonatype [WIP]

BenoitPerrot commented on algoliasearch-apache/pom.xml:
```diff
@@ -0,0 +1,51 @@
+            <groupId>com.algolia</groupId>
+            <artifactId>algoliasearch-core</artifactId>
+            <version>${algoliasearch-core-version}</version>
+            <type>test-jar</type>
```
add classifier too:
```xml
<classifier>tests</classifier>
```

Ant-hem replied:
Just curious, why do we need the `classifier` tag? 

When we run `mvn deploy` we already have a `.jar` suffixed with `tests` : `algoliasearch-apache-3.0.0-beta-1-tests.jar` Am I missing something?

Ant-hem replied:
Fixed!

BenoitPerrot commented on algoliasearch-apache/pom.xml:
```diff
@@ -0,0 +1,51 @@
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpasyncclient</artifactId>
+            <version>${async-http-client.version}</version>
```
I am not sure we really need a property for this version. We do not plan to allow its overriding on the command line, and the value is used only once (here).

Ant-hem replied:
Fixed!

BenoitPerrot commented on algoliasearch-apache/pom.xml:
```diff
@@ -0,0 +1,51 @@
+        <dependency>
+            <groupId>com.google.code.findbugs</groupId>
+            <artifactId>jsr305</artifactId>
+            <version>${jsr305.version}</version>
```
Remove the <version> tag; the version is already specified in the dependencyManagement section of the parent pom.

Ant-hem replied:
I moved findbugs in the parent pom as the library is shared with both children

BenoitPerrot commented on algoliasearch-apache/src/main/java/com/algolia/search/apache/ApacheAnalyticsClient.java:
```diff
@@ -0,0 +1,38 @@
+ * @see <a href="https://www.algolia.com/doc/rest-api/analytics/">Algolia.com</a>
+ */
+@SuppressWarnings("WeakerAccess")
+public class ApacheAnalyticsClient {
```
I suggest to drop the "Apache" prefix: the code is already in the com.algolia.search.**apache** package.

Ant-hem replied:
I don't have a strong opinion on it. Do you have example of other libraries in the Java world doing something similar? I couldn't find one.

BenoitPerrot replied:
Actually we don't see it because that is the default construction. Our own classes are for example "com.algolia.search.AccountClient", and not "com.algolia.search.AlgoliaSearchAccountClient"

It's the very purpose of _namespaces_: putting things in the same named set :P

Ant-hem replied:
Why not having all the `Apache-` classes in `com.algolia.search` then it's very clear for the documentation/customers how to instantiate a client?

I find this (in doc) clearer:
```java
SearchClient client = ApacheSearchClient.create(...);
```

than: 
```java
SearchClient client = com.algolia.search.apache.SearchClient.create(....);
```

ping @aseure

Ant-hem replied:
Or let's remane `ApacheSearchClient` by `DefaultSearchClient`

```Java
SearchClient client = DefaultSearchClient.create(...);
````

BenoitPerrot replied:
This sounds like a good approach to me 👍

BenoitPerrot commented on algoliasearch-apache/src/main/java/com/algolia/search/apache/ApacheHttpRequester.java:
```diff
@@ -36,18 +42,13 @@
   private final ConfigBase config;
 
-  public ApacheHttpRequester(@Nonnull ConfigBase config) {
+  ApacheHttpRequester(@Nonnull ConfigBase config) {
```
likewise: the Apache prefix could be dropped

BenoitPerrot commented on algoliasearch-apache/src/main/java/com/algolia/search/apache/ApacheInsightsClient.java:
```diff
@@ -0,0 +1,38 @@
+ * @see <a href="https://www.algolia.com/doc/rest-api/insights/">Algolia.com</a>
+ */
+@SuppressWarnings("WeakerAccess")
+public class ApacheInsightsClient {
```
likewise: the Apache prefix could be dropped

BenoitPerrot commented on algoliasearch-apache/src/main/java/com/algolia/search/apache/ApacheSearchClient.java:
```diff
@@ -0,0 +1,38 @@
+ * @see <a href="https://www.algolia.com/doc/rest-api/search/">Algolia.com</a>
+ */
+@SuppressWarnings("WeakerAccess")
+public class ApacheSearchClient {
```
likewise: the Apache prefix could be dropped

BenoitPerrot commented on algoliasearch-apache/pom.xml:
```diff
@@ -0,0 +1,51 @@
+
+    <properties>
+        <async-http-client.version>4.1.4</async-http-client.version>
+        <jsr305.version>3.0.2</jsr305.version>
```
remove this jsr305 property (see below why)

BenoitPerrot commented on algoliasearch-core/pom.xml:
```diff
@@ -0,0 +1,53 @@
+        <jackson.version>2.9.8</jackson.version>
+        <slf4j.version>1.7.25</slf4j.version>
+        <jsr305.version>3.0.2</jsr305.version>
+        <assertj-core.version>3.12.1</assertj-core.version>
```
Remove assertj-core.version: it is never used

Ant-hem replied:
Fixed

BenoitPerrot commented on algoliasearch-core/pom.xml:
```diff
@@ -0,0 +1,53 @@
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <jackson.version>2.9.8</jackson.version>
+        <slf4j.version>1.7.25</slf4j.version>
+        <jsr305.version>3.0.2</jsr305.version>
```
I am not sure properties are needed for "slf4j" and "jsr305": their values are used only once each

Ant-hem replied:
I was told it's always a good practice to have all the version in properties. Should I move them?

BenoitPerrot replied:
Your call. Do you remember the rationale of this best practice?

On my side I really do not see the point to keep them out-of-line if they are not to be overridden on the command line or in Maven, and if they are used only in one place.

Note that we can go like the current construction, and decide later on too.

BenoitPerrot commented on pom.xml:
```diff
@@ -1,315 +1,188 @@
+    <description>Java client for Algolia Search API</description>
+    <url>https://github.com/algolia/algoliasearch-client-java-2</url>
+
+    <!-- sub modules -->
```
remove this comment: the < modules > says it all :)

BenoitPerrot commented on src/main/java/com/algolia/search/models/indexing/Query.java:
```diff
@@ -521,6 +521,15 @@ public Query setEnablePersonalization(Boolean enablePersonalization) {
     return this;
   }
 
+  public Integer getPersonalizationImpact() {
```
blargh this change is completely unrelated to the multi-module thing :) you should push that in another PR

Ant-hem replied:
arg I checked out too fast 😞

BenoitPerrot commented on algoliasearch-core/src/main/java/com/algolia/search/util/AlgoliaUtils.java:
```diff
@@ -93,30 +89,6 @@ public static Boolean isNullOrEmptyWhiteSpace(final String stringToCheck) {
             + "must have an objectID property or a Jackson annotation @JsonProperty(\"objectID\")");
   }
 
-  public static CompletableFuture<HttpResponse> toCompletableFuture(
```
This inlining should be done in a separate, dedicated commit: it is actually unrelated to the move to multi-module

Ant-hem replied:
It is! This method takes an `org.apache.http.HttpResponse` as an input, as we wanted  `algoliasearch-core` to be completely independent from any http implementetion we had to move this helper in  `algoliasearch-apache - ApacheRequester`.

BenoitPerrot replied:
> It is!

It is not: it is possible to inline this helper in ApacheHttpRequester right now - without splitting the code into multi-module. For sure, it must be done for splitting cleanly. But it can be done beforehand.

The rationale is: separate concerns, keep commits atomic - always, always, always. So that if you have to revert a commit, it's less a big deal. And so that reviewing code is less stressful on the brain.

... Up to you! (seriously)

BenoitPerrot reviewed (changes requested):
- Please address the elements written in the imperative form ("Please do that" or "Remove this")
- Feel free not to follow my suggestions on the others (eg "I am unsure we need a property" or "We could drop the prefix 'Apache' ")

BenoitPerrot commented on pom.xml:
```diff
@@ -34,6 +34,14 @@
+            <name>algolia-maven</name>
+            <url>https://api.bintray.com/maven/algolia/maven/algoliasearch/;publish=1</url>
+        </repository>
+    </distributionManagement>
```
Let's do this in another PR: it is unrelated to switching to multi-modules, and it can break things badly, so let's keep the two separate (so if we need to revert, we can revert just the use of bintray)

aseure committed:
doc: fix typos in Javadoc comments

aseure committed:
chore: remove useless Maven plugins

aseure committed:
chore: set algolia package version in a single place (in parent pom.xml)

Ant-hem committed:
Updated: POM.xml with review

- Added: classifier tests in algoliasearch-test
- Removed apache version in properties
- Removed unused assertj-core version in algoliasearch-core
- Moved findbugs in parent pom
- Moved project encoding in parent pom

aseure committed:
chore: add versions-maven-plugin to handle version changes and update release.sh script

Ant-hem committed:
Renamed: ApacheClients to DefaultClients

aseure committed:
chore: stop using versions variable in pom.xml files in favor of versions-maven-plugin set command

aseure committed:
chore: fix pom.xml files according to Maven warnings

aseure committed:
chore: ensure deploy was effectively done before git commit/tag/push in release.sh

BenoitPerrot commented on algoliasearch-apache/src/main/java/com/algolia/search/DefaultSearchClient.java:
```diff
@@ -0,0 +1,36 @@
+ * @see <a href="https://www.algolia.com/doc/rest-api/search/">Algolia.com</a>
+ */
+@SuppressWarnings("WeakerAccess")
+public class DefaultSearchClient {
```
(yes I definitely like it)
